content/data-sharing/persistent-data-structures.md

```diff
@@ -0,0 +1,72 @@
+
+## Solution
+
+```pony
+use "collections/persistent"
+
+actor Collector
+  """
+  Receives characters via it's `collect` behavior and stores them.
+  Every 10 characters we receive results in the entire array being sent all to
+  the receiver.
+  """
+  let _receiver: Receiver
+  var _data: List[U8] = Lists[U8].empty()
+
+  new create(receiver: Receiver) =>
+    _receiver = receiver
+
+  be collect(char: U8) =>
+    _data = _data.prepend(char)
+
+  be send(to: Receiver) =>
+    to.receive(_data)
+
+actor Receiver
+  """
+  Receives an array of characters from a collector
+  """
+  be receive(data: List[U8]) =>
+    // do something with `data`
+    None
+```
+
+## Discussion
+
+That's a pretty simple looking solution, especially when you compare it to the [copying pattern](/data-sharing/copying.html), which is another way to solve this problem. So what's going on here?
+
+The key is that our `_data` List isn't mutable. In fact, `var _data: Lists[U8] = Lists[U8].empty()` is creating a `val`. So, we aren't dealing with mutable data; we are creating a new immutable List each time we "mutate" it. That's why we need to assign to `_data` in the `collect` method:
+
+```pony
+  be collect(char: U8) =>
+    _data = _data.prepend(char)
+```
+
+And it's also why we defined `_data` using a `var` rather than a `let` binding. Each time we update, we are creating a new List and binding our `_data` variable to the new list.
+
+Because `_data` is a `val`, it's already safe to share between actors. We don't need to do anything special with it when we want to send it to another actor:
+
+```pony
+  be send(to: Receiver) =>
+    to.receive(_data)
+```
+
+Persistent data structures go hand in hand with the [copying pattern](/data-sharing/copying.html) as a means of sharing mutable data between actors. Each method involves copying data. Which pattern should you pick? Whichever one will minimize the number of copies.
+
+Persistent data structures involve copying on each update. The copying pattern involves a copy each time we send the data to another actor. As a general rule of thumb, figure out which you will do more: update or send. If you are updating more, use the copying pattern. If you are sending more, use persistent data structures.
+
+In the end, that's just a rule of thumb. Your best bet is to benchmark and pick the method that gives you the best performance for your use case.
+
+If you aren't familiar with persistent data structures, we suggest you pick up a copy of the book [Purely Functional Data Structures](https://www.amazon.com/Purely-Functional-Data-Structures-Okasaki/dp/0521663504). Another option is to download a copy of [the thesis](https://www.cs.cmu.edu/~rwh/theses/okasaki.pdf) upon which the book is based. The Pony [standard library](https://stdlib.ponylang.io/collections-persistent--index/) contains a few, but you may need to design your own.
```

Comment:
It may also be worth mentioning these two blog posts that explain much of the detail behind the persistent `Vec` and `HashMap` that are in the pony standard library:
- https://hypirion.com/musings/understanding-persistent-vector-pt-1
- https://hypirion.com/musings/persistent-vector-performance-summarised

Okasaki's work is definitely one of the best references for persistent/functional data structures in general though.

Reply:
Will add.